tika-parsers/src/main/java/org/apache/tika/parser/html/BoilerpipeContentHandler.java

```diff
@@ -230,14 +229,20 @@ public void endDocument() throws SAXException {
                     case CONTINUE:
                         // Now emit characters that are valid. Note that boilerpipe pre-increments the character index, so
                         // we have to follow suit.
-                        for (char[] chars : element.getCharacters()) {
+                        for (int i = 0; i < element.getCharacters().size(); i++) {
+                            char[] chars = element.getCharacters().get(i);
                             curCharsIndex++;
+                            boolean isValidCharacterRun = validCharacterRuns.get(curCharsIndex);
 
-                            if (validCharacterRuns.get(curCharsIndex)) {
+                            // https://issues.apache.org/jira/projects/TIKA/issues/TIKA-2683
+                            // Allow exempted characters to be written
+                            if (isValidCharacterRun ||
+                                    (chars.length == 1 && whitelistCharSet.contains(chars[0]))) {
                                 delegate.characters(chars, 0, chars.length);
 
                                 // https://issues.apache.org/jira/browse/TIKA-961
-                                if (!Character.isWhitespace(chars[chars.length - 1])) {
+                                if (isValidCharacterRun && i == element.getCharacters().size() - 1
```

Comment:
Could you explain why we now want to check for `i == element.getCharacters().size() - 1`, where that wasn't previously done by the patch for TIKA-961?

Reply:
It seems `element.getCharacters()` returns a section which corresponds to a tag in the HTML web page. Inside that there are sub-sections created by char length or other heuristics based on the BoilerPipe algorithm. 

This check ensures that we are detecting/adding whitespace characters at the end of section and not in between. The current behavior doesn't take care of this and you may see improper new-line characters between the text (Example - https://en.wikipedia.org/wiki/Blobfish)